chainerx_cc/chainerx/index_iterator.h

```diff
@@ -64,10 +64,43 @@ class IndexIterator {
     CHAINERX_HOST_DEVICE void Set(int64_t i) {
         CHAINERX_ASSERT(total_size_ > 0);
         raw_index_ = i;
+#ifdef __CUDA_ARCH__
+        if (total_size_ >= 1LL << 32) {
+            // 64-bit division is very slow on GPU
+            uint64_t a = static_cast<uint64_t>(i);
+            for (int8_t dim = kNdim; --dim > 0;) {
+                uint64_t s = static_cast<uint64_t>(shape_[dim]);
+                if (s & (s - 1)) {
+                    uint64_t t = a / s;
+                    index_[dim] = a - t * s;
```

Comment:
Cast is needed here too.
```suggestion
                    index_[dim] = static_cast<int64_t>(a - t * s);
```